### PR TITLE
Adds "name" property to Rule and RuleResult

### LINKED ISSUE #144: Feature Request - Naming Rules for RuleResult Output
My application has the need to save the result of every rule.  This includes things like rule logic, input fact values, and the result of each rule - all things that are included out of the box in this library with RuleResult 👍 .

However, I also need to record the name of the rule that fired.  This is a concept that exists only to uniquely identify rules stored in my DB and in our records of results, but is largely irrelevant to the execution engine.

I was wondering if this would be a feature you'd be willing to support:
1. An optional field of `id` or `name` when defining a rule:
```javascript
ruleOptions = {
  id: 'Rule1',
  conditions: {
    all: [
      {
        fact: 'my-fact',
        operator: 'equal',
        value: 'some-value'
      }
    ]
  },
  event: {
    type: 'my-event',
    params: {
      customProperty: 'customValue'
    }
  }
}
```
2. Output this in `RuleResult`:
```javascript
{
  id: 'Rule1'
  result: false,
  conditions: {
    all: [
      {
        fact: 'my-fact',
        operator: 'equal',
        value: 'some-value',
        result: false,
        factResult: 'other-value'
      }
    ]
  },
  event: {
    type: 'my-event',
    params: {
      customProperty: 'customValue'
    }
  },
  priority: 1,
}
```

### PR DESCRIPTION
**Pull request by MattCopenhaver**
Resolves #144 

I'm not sure that `name` is best here.  Maybe `id` or something similar would be better.  I didn't go with `id`, as that would insinuate that it must be unique per rule, which is not the case.

### CODE REVIEW

File: src/rule.js [hunk 1]
```diff
@@ -49,6 +53,18 @@ class Rule extends EventEmitter {
     return this
   }
 
+  /**
+   * Sets the name of the rule
+   * @param {string} name - only non-empty strings are allowed
+   */
+  setName (name) {
+    if (!name || typeof name !== 'string') {
```
Comment (CacheControl) [inline src/rule.js]:
```suggestion
    if (!name) {
```

@MattCopenhaver I'm wondering if there might be some benefit in allowing this field to be multiple types. For example, if a consumer wanted to set `name` to be an js object to enable multiple properties, e.g.
```js
rule.setName({
 id: '1234',
 type: 'user'
})
```

or in the case of multiple identifiers:
```js
rule.setName({
 pk: 1234,
 uuid: 'f6f14d07-04a4-4db1-9936-75b61069f25b'
})
```

definitely both are edge cases, but if the functionality comes for 'free' by being less strict, I don't see a down side to enabling it.

Comment (MattCopenhaver) [reply src/rule.js]: Good suggestion.  I'll make this change.

**Comment by MattCopenhaver** [comment]
Thanks for the quick response @CacheControl!  I'll make the requested changes when I get home tonight.

File: src/rule.js (hunk 1 above)
Comment (CacheControl) [reply src/rule.js]: I went back and forth on the `id`/`name`/? thing as well. Honestly I think we should just stick with `name` - like you say the uniqueness isn't being enforced by the library, and these additional cases using objects will be an edge case.

**Comment by MattCopenhaver** [comment]
@CacheControl ready for another look.  Excuse my file by file commits (you don't even want to know)...

**Comment by CacheControl** [comment]
Looks great! Thanks for the contribution @MattCopenhaver. This is published as **3.1.0**

**Comment by MattCopenhaver** [comment]
Awesome, thanks!

**Pull request merged**